src/basicfuns.jl

```diff
@@ -246,6 +249,66 @@ function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     end
 end
 
+"""
+    logsumexp_onepass(X)
+
+Compute `log(sum(exp, X))` in a numerically stable way that avoids intermediate under- and
+overflow.
+
+In contrast to [`logsumexp`](@ref) the result is computed using a single pass over the data.
+`X` should be an iterator of real numbers.
+
+# References
+
+[Sebastian Nowozin: Streaming Log-sum-exp Computation.](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)
+"""
+function logsumexp_onepass(X)
+    # fallback for empty collections
+    isempty(X) && return log(sum(X))
+
+    # perform single pass over the data
+    xmax, r = _logsumexp_onepass(X, Base.IteratorEltype(X))
+
+    return xmax + log(r)
+end
+
+# with initial element: required by CUDA
```

Comment:
I guess it's also required by `Array` when the input is empty, right?

Reply:
No, empty inputs are handled by `log(sum(X))`, as before. Actually we never want to use this method apart for GPU arrays - the type calculations are doomed to fail e.g. for not concretely typed arrays and hence it is much safer to not rely on them and instead just compute the output. I'm not sure why the type calculations in https://github.com/JuliaGPU/GPUArrays.jl/blob/356c5fe3f83f76f8139b4edf5536cbd08d48da7f/src/host/mapreduce.jl#L49-L52 fail, I can check if this can be fixed from our side. Alternatively, maybe another package such as NNlib should provide a GPU-compatible overload of StatsFuns.logsumexp?

Reply:
OK. They fail because `neutral_element` is only defined for a few particular functions. Another approach would be to use the type of the first element, since we already handle the case where the array is empty. That way we wouldn't need two different methods depending on whether the eltype is known.

Given that it seems doable, it would be nice to have something that works for all arrays without another package having to overload the method.

Reply:
Runs on the GPU fail even before this point due to the type calculations. IMO it is just bad that GPUArrays requires us to provide an initial or neutral element. We really never want to do that, regardless if there is a first element or not. Retrieving the first element is bad for any stateful iterator, and is a scalar indexing operation on the GPU (which leads to warnings when calling the function - and we can not remove them without depending on CUDA). Additionally, the docstring of `mapreduce` says that
```julia
It is unspecified whether init is used for non-empty collections.
```
so there is really no point in providing `init` here - apart from GPU compatibility. The problem is that we can not restrict our implementation with the `init` hack to only GPU arrays without taking a dependency on CUDA. Hence I don't think it is completely unreasonable to provide a GPU-compatible implementation somewhere else. An alternative would be to not use the one-pass algorithm as the default algorithm (for arrays).

Reply:
Retrieving the first element should be OK for any `AbstractArray` (contrary to some iterables). It's too bad that CUDA prints warnings in that case.

Let's keep what you have now: since the method only accepts arrays with `T<:Real` element type, `float(T)` will (almost) certainly work.